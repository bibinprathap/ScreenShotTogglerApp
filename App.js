/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React, {useRef, useState, useEffect} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  View,
  Text,
  StatusBar,
  NativeModules,
  TouchableOpacity,
  Button,
  ActivityIndicator,
  AppState,
  Platform,
  Alert,
} from 'react-native';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';

import {
  Header,
  LearnMoreLinks,
  Colors,
  DebugInstructions,
  ReloadInstructions,
} from 'react-native/Libraries/NewAppScreen';

import DeviceInfo from 'react-native-device-info';
import GetLocation from 'react-native-get-location';
import {NetworkInfo} from 'react-native-network-info';

const {ScreenShot} = NativeModules;

const SecurityScreen = () => <View />;

const showSecurityScreenFromAppState = (appState) =>
  ['background', 'inactive'].includes(appState);

const App: () => React$Node = () => {
  const [disableScreenShotActive, setDisableScreenShotActive] = useState(false);
  const [active, setActive] = useState(true);
  const [postingDeviceinfo, setPostingDeviceinfo] = useState(false);
  const appState = useRef(AppState.currentState);
  const [appStateVisible, setAppStateVisible] = useState(appState.current);

  useEffect(() => {
    AppState.addEventListener('change', handleAppStateChange);
    return () => {
      AppState.removeEventListener('change', handleAppStateChange);
    };
  }, []);

  const handleAppStateChange = (nextAppState) => {
    if (
      appState.current.match(/inactive|background/) &&
      nextAppState === 'active'
    ) {
      console.log('App has come to the foreground!');
    }

    appState.current = nextAppState;
    setAppStateVisible(appState.current);
    console.log('AppState', appState.current);
  };
  const sendDeviceinfotoServer = async (nextAppState) => {
    setPostingDeviceinfo(true);

    const ipAddressPromise = new Promise((resolve, reject) => {
      NetworkInfo.getIPV4Address().then((ipv4Address) => {
        console.log(ipv4Address);
        resolve(ipv4Address);
      });
    });
    const macAddressPromise = new Promise((resolve, reject) => {
      DeviceInfo.getMacAddress().then((mac) => {
        resolve(mac);
      });
    });

    const locationPromise = new Promise((resolve, reject) => {
      GetLocation.getCurrentPosition({
        enableHighAccuracy: true,
        timeout: 15000,
      })
        .then((location) => {
          console.log(location);
          resolve(location);
        })
        .catch((error) => {
          // const {code, message} = error;
          // console.warn(code, message);
          reject(error);
        });
    });

    const screenshotstatusPromise = Promise.resolve(disableScreenShotActive);
    const osName = DeviceInfo.getSystemName();
    const osNamePromise = Promise.resolve(osName);

    Promise.all([
      macAddressPromise,
      locationPromise,
      ipAddressPromise,
      screenshotstatusPromise,
      osNamePromise,
    ]).then((values) => {
      console.log(values);
      submitDeviceInfo({
        ipv4Address: values[2],
        location: values[1],
        osName: values[4],
        screenshotstatus: values[3],
        getMacAddress: values[0],
      });
    });
  };
  const submitDeviceInfo = async ({
    ipv4Address,
    location,
    osName,
    screenshotstatus,
    getMacAddress,
  }) => {
    const postData = {
      ipv4Address,
      location,
      osName,
      screenshotstatus,
      getMacAddress,
    };
    try {
      let response = await fetch('http://localhost:5000/api/users', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(postData),
      });
      response = await response.json();
      setPostingDeviceinfo(false);
    } catch (err) {
      setPostingDeviceinfo(false);
      console.log('Error:', err);
    }
  };

  const renderApp = (
    <>
      <SafeAreaView style={{flex: 1}}>
        <View style={styles.container}>
          <View style={{flex: 2}}></View>
          <View style={{paddingVertical: 25, flex: 2}}>
            <View
              style={{
                width: 120,
                height: 120,
                backgroundColor: 'red',
                borderRadius: 8,
              }}></View>
          </View>
          <View style={{flex: 2}}>
            <Text>Enable or disable ScreenShot</Text>
            {/* <Button
              onPress={() => {
                if (Platform.OS === 'android') {
                  ScreenShot.start().then(() => {

                  });
                }
                sendDeviceinfotoServer();
                setDisableScreenShotActive(true);
                Alert.alert('ScreenShot Disabled');
              }}
              title="Activate"
              color="red"
              accessibilityLabel="Click this button to   disable ScreenShot"
            /> */}

            <TouchableOpacity
              activeOpacity={0.8}
              testID={'submit-button'}
              style={{justifyContent: 'center', alignItems: 'center'}}
              onPress={() => {
                if (!active) {
                  if (Platform.OS === 'android') {
                    ScreenShot.stop().then(() => {});
                  }

                  setDisableScreenShotActive(false);
                  setActive(true);
                  sendDeviceinfotoServer();
                  Alert.alert('ScreenShot Enabled');
                } else {
                  if (Platform.OS === 'android') {
                    ScreenShot.start().then(() => {});
                  }
                  sendDeviceinfotoServer();
                  setDisableScreenShotActive(true);
                  setActive(false);
                  Alert.alert('ScreenShot Disabled');
                }
              }}>
              <View
                style={[
                  styles.signUp,
                  {backgroundColor: active ? '#00cc66' : '#3333cc'},
                ]}>
                <FontAwesome5
                  name={active ? 'times' : 'check-circle'}
                  solid
                  size={18}
                  color="white"
                />
                {/* <ActivityIndicator size="small" color="white" /> */}
                <Text style={styles.span3}>
                  {active ? 'DE-ACTIVATE' : 'ACTIVATE'}
                </Text>
              </View>
            </TouchableOpacity>
            {/* <Button
              onPress={() => {
                if (Platform.OS === 'android') {
                  ScreenShot.stop().then(() => {

                  });
                }

                setDisableScreenShotActive(false);
                sendDeviceinfotoServer();
                Alert.alert('ScreenShot Enabled');
              }}
              title="Deactivate"
              color="green"
              accessibilityLabel="Click this button to Enable   ScreenShot"
            /> */}
          </View>
        </View>
      </SafeAreaView>
    </>
  );
  //
  return Platform.OS === 'ios' &&
    disableScreenShotActive &&
    showSecurityScreenFromAppState(appStateVisible) ? (
    <SecurityScreen />
  ) : (
    renderApp
  );
};

const styles = StyleSheet.create({
  scrollView: {
    backgroundColor: Colors.lighter,
  },
  button: {
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 10,
    borderRadius: 20,
    marginVertical: 10,
    color: 'white',
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',
  },
  engine: {
    position: 'absolute',
    right: 0,
  },
  body: {
    backgroundColor: Colors.white,
  },
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  signUp: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#08945C',
    paddingLeft: 25,
    paddingRight: 25,
    height: 40,
    width: 150,
    borderRadius: 30,
    marginVertical: 16,
  },
  span3: {
    color: 'white',
    fontSize: 12,
    paddingLeft: 20,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
    color: Colors.black,
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
    color: Colors.dark,
  },
  highlight: {
    fontWeight: '700',
  },
  footer: {
    color: Colors.dark,
    fontSize: 12,
    fontWeight: '600',
    padding: 4,
    paddingRight: 12,
    textAlign: 'right',
  },
});

export default App;
