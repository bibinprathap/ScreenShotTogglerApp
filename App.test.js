/**
 * @format
 */

import {NativeModules, Platform, AppState, Alert} from 'react-native';
import React from 'react';
import App from './App';
import {render, cleanup, fireEvent} from '@testing-library/react-native';

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native'); // use original implementation, which comes with mocks out of the box

  // mock modules/components created by assigning to NativeModules
  RN.NativeModules.ScreenShot = {
    start: jest.fn(() => Promise.resolve('1.0')),
    stop: jest.fn(() => Promise.resolve('1.0')),
  };
  RN.NativeModules.RNCNetInfo = {
    getCurrentConnectivity: jest.fn(),
    isConnectionMetered: jest.fn(),
    addListener: jest.fn(),
    removeListeners: jest.fn(),
  };

  return RN;
});
//, {ValidationErrors}
jest.mock('react-native/Libraries/Alert/Alert', () => ({
  alert: jest.fn(),

  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
}));
// jest.mock('NativeModules', () => {
//   return {
//     addEventListener: jest.fn(),
//     removeEventListener: jest.fn(),
//     inFocusDisplaying: jest.fn(),
//     ScreenShot: jest.fn(),
//   };
// });
jest.mock('react-native-device-info', () => {
  return {
    getMacAddress: jest.fn(() => Promise.resolve('1.0')),
    getSystemName: jest.fn(() => Promise.resolve('My App')),
    getVersion: () => 4,
  };
});

describe('Screenshot Toggler', () => {
  let wrapper;
  beforeEach(() => {
    Platform.OS = 'android';

    AppState.currentState = 'active';
    wrapper = render(<App />);
  });

  afterEach(() => {
    cleanup();
    wrapper = null;
    Platform.OS = 'android';
  });

  it('User is able to see Logo and Activate Button', () => {
    expect(wrapper).toMatchSnapshot();
    const {getByTestId} = wrapper;
    const submitButton = getByTestId('submit-button');
    expect(submitButton).toBeEnabled();
  });

  it('user Pressed activate button', () => {
    const {getByTestId} = wrapper;
    const submitButton = getByTestId('submit-button');
    fireEvent(submitButton, 'onPress');
    //const validationError = getByTestId('text-error');

    // expect(validationError).toBeTruthy();
    // expect(validationError.props.children).toBe(ValidationErrors.Activated);
  });

  it('user Pressed deactivate button', () => {
    const {getByTestId} = wrapper;
    const submitButton = getByTestId('submit-button');
    const submitButton2 = getByTestId('submit-button');
    fireEvent(submitButton, 'onPress');
    fireEvent(submitButton2, 'onPress');
    //const validationError = getByTestId('text-error');

    // expect(validationError).toBeTruthy();
    // expect(validationError.props.children).toBe(ValidationErrors.Deactivated);
  });
});
