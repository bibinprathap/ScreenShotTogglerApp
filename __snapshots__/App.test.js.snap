// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot Toggler User is able to see Logo and Activate Button 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 2,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 2,
          "paddingVertical": 25,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "red",
            "borderRadius": 8,
            "height": 120,
            "width": 120,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 2,
        }
      }
    >
      <Text>
        Enable or disable ScreenShot
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID="submit-button"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#08945C",
                "borderRadius": 30,
                "display": "flex",
                "flexDirection": "row",
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 16,
                "paddingLeft": 25,
                "paddingRight": 25,
                "width": 150,
              },
              Object {
                "backgroundColor": "#00cc66",
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 18,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome5Free-Solid",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "900",
                },
              ]
            }
          >
            ÔÄç
          </Text>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 12,
                "paddingLeft": 20,
              }
            }
          >
            DE-ACTIVATE
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
